File: backend/scripts/capture_housing_storage_state.py
import argparse
import asyncio
from pathlib import Path

from playwright.async_api import async_playwright


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Capture Housing.com Playwright storage state from a human-verified browser session.")
    parser.add_argument(
        "--output",
        default="backend/scripts/housing_storage_state.json",
        help="Where to save Playwright storage state JSON.",
    )
    parser.add_argument(
        "--url",
        default="https://housing.com/rent/flats-for-rent-in-kolkata-west-bengal-P40qcmycif4m431jo",
        help="Housing URL to open before saving state.",
    )
    return parser.parse_args()


async def main() -> None:
    args = parse_args()
    output_path = Path(args.output).resolve()
    output_path.parent.mkdir(parents=True, exist_ok=True)

    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=False, args=["--disable-blink-features=AutomationControlled"])
        context = await browser.new_context(locale="en-IN", timezone_id="Asia/Kolkata")
        await context.add_init_script(
            """
            Object.defineProperty(navigator, 'webdriver', {
                get: () => undefined,
            });
            """
        )
        page = await context.new_page()
        await page.goto(args.url, wait_until="domcontentloaded", timeout=90000)

        print("Browser opened.")
        print("Complete any challenge/login manually in the browser window.")
        print("When done, return to this terminal and press Enter to save storage state.")
        input()

        await context.storage_state(path=str(output_path))
        await context.close()
        await browser.close()

    print(f"Saved storage state to: {output_path}")


if __name__ == "__main__":
    asyncio.run(main())

File: backend/scripts/captureHousingStorageState.ts
import { Command } from "commander";
import fs from "fs";
import path from "path";
import { chromium } from "playwright";
import readline from "readline";

async function main() {
  const program = new Command();
  program
    .description("Capture Housing.com Playwright storage state from a human session.")
    .option("--output <path>", "Where to save Playwright storage state JSON", "scripts/housing_storage_state.json")
    .option(
      "--url <url>",
      "Housing URL to open",
      "https://housing.com/rent/flats-for-rent-in-kolkata-west-bengal-P40qcmycif4m431jo"
    );

  program.parse();
  const options = program.opts();

  const outputPath = path.resolve(options.output);
  fs.mkdirSync(path.dirname(outputPath), { recursive: true });

  const browser = await chromium.launch({
    headless: false,
    args: ["--disable-blink-features=AutomationControlled"],
  });

  const context = await browser.newContext({
    locale: "en-IN",
    timezoneId: "Asia/Kolkata",
  });

  await context.addInitScript(`
    Object.defineProperty(navigator, 'webdriver', {
      get: () => undefined,
    });
  `);

  const page = await context.newPage();
  await page.goto(options.url, { waitUntil: "domcontentloaded", timeout: 90000 });

  console.log("Browser opened.");
  console.log("Complete any challenge/login manually in the browser window.");
  console.log("When done, return to this terminal and press Enter to save storage state.");

  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  await new Promise((resolve) => rl.question("", resolve));
  rl.close();

  await context.storageState({ path: outputPath });
  await context.close();
  await browser.close();

  console.log(`Saved storage state to: ${outputPath}`);
}

main().catch((err) => {
  console.error("Failed to capture storage state:", err);
  process.exit(1);
});
